src/compiler/checker.ts

```diff
@@ -27789,7 +27844,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
             }
         }
-        return changed ? mappedTypes && getUnionType(mappedTypes, noReductions ? UnionReduction.None : UnionReduction.Literal) : type;
+        return changed
+            ? mappedTypes &&
+                (toIntersection

```

Comment:
I don't like the inlining idea very much, I think it makes it less clear what's going on in `getConditionalTypeInstantiation`. Also, `mapType` is recursive. 

`mapTypeIntersection` would essentially be a copy of 90% of `mapType`, except for the `changed` variable and parameter `noReductions`, which wouldn't be needed. I am not opposed to this, because I think it is unlikely that we change the core logic of mapping over a type.

Refactoring the shared code into `mapTypeWorker` is kinda weird because of the special case for `changed`: in addition to returning `Type | Type[] | undefined`, `mapTypeWorker` would have to also return this `changed` flag so that the regular `mapType` can make an optimization. This actually made me realize there was a latent bug in my update: even if no type in the union type has changed after mapping, if `toIntersection` is true, we can't simply return the union type, but we have to convert it to an intersection. This doesn't matter for the single use of the function, because the types are guaranteed to change.

Let me know if I should keep the extra parameter to `mapType` or make a new `mapTypeToIntersection`, I don't have strong feelings either way.

Reply:
I guess I slightly prefer `mapTypeToIntersection`, although it's too bad that the return type `Type | IntersectionType` is nonsensical. But it does capture the difference between `mapType` returning `Type | UnionType` (also nonsensical).